### ISSUE TITLE
SDP constraint not satisfied 

### ISSUE THREAD

**Issue by osjacky430** [JuliaOpt/Convex.jl#266]
Hi, I'm trying to implement the simulation in the paper, and I've encountered the following convex problem

![image](https://user-images.githubusercontent.com/11375975/50851449-d67d7580-13b7-11e9-99db-2781486e9783.png)

which is a SDP constraint, where 

Q_i is a n x n, real, and semidefinite matrix
r_i is a n x 1 complex matrix
s_i, d_i are scalars, and d_i is non-negative

The code is 
```
using LinearAlgebra
using Distributions
using Random
using Convex
using SCS
using Mosek
using Printf

Nt = K = 3;
rng = MersenneTwister(1234);
chiSqDis = Chisq(2 * Nt);

S = ntuple(x->Variable(3, 3), K);

γ = 10^(1/10); # linear power ratio, 1dB
σ = 0.1;
C = Diagonal(fill(0.002, Nt)); # CSI error
ρ = 0.1;

ϵ = 1e-6;
feasible_num = 0;
realization = 2;

for iter = 1:realization
    global feasible_num;
    problem = minimize(tr(sum(S)));

    t = [];
    hBars = [];

    for i = 1:K
        hBar = randn(rng, ComplexF64, (Nt, 1));
        middle = S[i]/γ - sum(S[1:end .!= i]);

        Qi = sqrt(C) * middle * sqrt(C);
        ri = sqrt(C) * middle * hBar;
        si = real(hBar' * middle * hBar) - σ;
        di = sqrt(quantile(chiSqDis, 1 - ρ) / 2.0);
        ti = Variable();
        eye = Diagonal(fill(1, Nt));

        push!(t, ti);
        push!(hBars, hBar);

        problem.constraints += (S[i] == transpose(S[i]))
        problem.constraints += (S[i] ⪰ 0);
        problem.constraints += (ti ≥ 0);
        problem.constraints += ([Qi + ti * eye  ri; ri'   si - ti * di^2] ⪰ 0);
    end

    solve!(problem, MosekSolver());

    if problem.status == :Optimal
        feasible_num += 1;

        for j = 1:K
            global S;

            SumJ = S[j].value / γ - sum(Si.value for Si in S[1:end .!= j]);
            Qj = sqrt(C) * SumJ * sqrt(C);
            rj = sqrt(C) * SumJ * hBars[j];
            sj = real((hBars[j])' * SumJ * hBars[j])[1] - σ;
            tj = t[j].value;
            dj = sqrt(quantile(chiSqDis, 1 - ρ) / 2.0);

            eye = Diagonal(fill(1, Nt));

            println(eigvals([Qj + tj * eye rj; rj' sj - tj * dj^2]))
        end
    end
end

@printf("feasble rate: %f", feasible_num / convert(Float64, realization) * 100);

```

Despite the status is optimal, checking the eigenvalues of the [Qi + ti * eye  ri; ri'   si - ti * di^2], I found that all of them have a negative eigenvalue, which violates the constraint. 

Also, I've tried using ``` lambdamin([Qi + ti * eye  ri; ri'   si - ti * di^2]) >= 0``` still gives me negative eigenvalue. No idea what is going on now.

Btw, I'm using Mosek to solve the problem. Not sure if it helps.

Thanks for any advice and idea, really appreciate it.

[UPDATE] The full code is pasted, since it's hard to know whether it's my problem or not only by the given information.

**Comment by ericphanson** [comment]
I also ran into the second problem, the error 
```
Objective can not be a complex expression
```
when taking the trace of Hermitian PSD matrices. I tried to figure out how to fix it in the code, but didn't see a quick fix since the library doesn't seem to keep track of PSD constraints in a way that `diag` can know if it's argument is PSD (since trace is implemented as `tr(X) = sum(diag(X))`). Specifically, we would want

```julia
function sign(x::DiagAtom)
    return sign(x.children[1])
end
```
to be able to check if it's argument is PSD (or Hermitian even), but I couldn't see how to do that. I tried looking at some of the other disciplined convex programming libraries for other languages to see how they handle it, but I guess I'm spoiled by Julia because they all seem like a bit of a mess to me. But the easiest fix works fine-- just write it as
```julia
problem = minimize(real(tr(X)))
```
instead. Maybe that workaround should be added to the docs in lieu of a proper fix?
